veneer documentation (alpha) (#245)
operator-framework/olm-docs: content/en/docs/Reference/veneers.md

## Benefits

- A set of consistent expectations from including as standard functionality with `opm`
- A good set of use-cases for simplifying the ability of operator authors to manage FBCs, including contributing them to an index
- A transitive benefit from consistency of practice as authors move from operator to operator

The veneers which accompany `opm` are:
- [`basic veneer`](#basic-veneer), which provides simplification of FBC concepts; and
- [`semver veneer`](#semver-veneer), which provides the capability of sophisticated channel generation

## Basic Veneer
The `basic veneer` is an input schema which eliminates FBC information that can be retrieved from existing registries when we process it.
Users provide all components of an [FBC schema](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#olm-defined-schemas), but supply only the bundle image reference for any `olm.bundle` objects.  This results in a greatly simplified, smaller document.

In a very simple example, we define an `olm.package` and an `olm.channel` composed of two `olm.bundle` objects that have an image name attribute but no other attributes/properties.
```yaml

> ---

Reviewer: I feel that there may be a way to shorten naturally. Do we need all of the information here? Maybe a Hugo component can let us put this behind a drop down like Github does with `<details>` but I'm not super familiar with framework's limitations.

---

> title: "Veneers"

Reviewer 2: ```suggestion
title: "File-based catalog veneers"
```

linkTitle: "Veneers"
weight: 5
date: 2022-06-30
---

File-Based Catalogs (FBC) are a major improvement to the imperative update graph approaches of previous versions. FBCs give operator authors a [declarative and deterministic approach to defining their update graph](https://olm.operatorframework.io/docs/concepts/olm-architecture/operator-catalog/creating-an-update-graph/). However, FBCs can get complex, especially as the number of releases and dependencies scale.

> 

Reviewer 3: This guide is an in-depth walkthrough to let you know how to work with veneers. Veeners are APIs which will help Operator authors and Index catalog admins generate and keep maintained their content using the [FBC][] format.

@grokspawn IHMO we need to try to be more direct here. Who is the audience? Why they will look at this doc? We need to make the info clear and less verbose as possible for who use it and try to abstract complexity that only matters to those keep it maintained.

Example: `Enter the `veneer`, a simpler transitional step to generating an FBC.` 

What means a transitional step to generate an FBC for those that are just looking to publish their solutions in OperatorHub io or OCP catalogs?

> File-Based Catalogs (FBC) are a major improvement to the imperative update graph approaches of previous versions.  FBCs give operator authors a [declarative and deterministic approach to defining their update graph](https://olm.operatorframework.io/docs/concepts/olm-architecture/operator-catalog/creating-an-update-graph/).  Unfortunately, FBCs can get complex, especially as the number of releases and dependencies scale.

Reviewer 4: There are a lot of double-white-spaces at the beginning of every sentence in this doc. Not sure why. Is this for visual enhancement? Otherwise, please clear them up as it creates strange spacing while reading the doc.

Reviewer 4: You will notice the larger space at the beginning of every sentence. It messes up the even spacing between every word. Unless there is a reason behind that extra space, I would recommend removing it.

> Enter the `veneer`, a simpler transitional step to generating an FBC.

Reviewer 4: you mean tool or step?

> ## Design

Reviewer 3: Why does the Design matter for those that will use it? 
Is not it defined in a PE for those who keep the solution maintained?

There are two components to every `veneer`:
1. An arbitrary API

> 2. An executable which processes #1 and produces a valid FBC.

Reviewer 5: It looks like the preview isn't rendering these numbered bullet points. You may need to use bullet points throughout this entire document.

Author: I checked the HTML output, and it is correctly using \<ol\>...\</ol\> tags:
```html
 <p>There are two components to every <code>veneer</code>:</p>
<ol>
<li>An arbitrary API</li>
<li>An executable which processes #1 and produces a valid FBC.</li>
</ol>
```
so this looks like just a preview problem.

## Benefits

- A set of consistent expectations from including as standard functionality with `opm`
- A good set of use-cases for simplifying the ability of operator authors to manage FBCs, including contributing them to an index

> - A transitive benefit from consistency of practice as authors move from operator to operator

Reviewer 3: Would we not put the info in a more friendly ( end-user way)?

Example: When should I use it?

The veneers which accompany `opm` are:
- [`basic veneer`](#basic-veneer), which provides simplification of FBC concepts; and

> - [`semver veneer`](#semver-veneer), which provides the capability of sophisticated channel generation

Reviewer 3: In this case, I understand that we do not need to config replace/skips/skipRange. Am I right

The veneer will create the upgrade graph based on the server of the bundles.
On top of that, it also manages the channel creation based on the server which makes things eaiser for authors does not need to configure them. 

So, should not batter the small description say that this veener create the upgrade graph and channels using the bundle versions follow the semver convention and link here the convention. Wdyt?

## Basic Veneer

> The `basic veneer` is an input schema which eliminates FBC information that can be retrieved from existing registries when we process it.

Reviewer 3: > The `basic veneer` is an input schema which eliminates FBC information that can be retrieved from existing registries when we process it.

How this phase can be clear for those that are only producing solutions or maintaining an index?
Indeed what exactly does that mean?

Author: The next sentence says: 
> Users provide all components of an [FBC schema](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#olm-defined-schemas), but supply only the bundle image reference for any olm.bundle objects. This results in a greatly simplified, smaller document.

Does that help?

I'm using the [general-to-specific](https://www.thoughtco.com/general-to-specific-order-composition-1690812) approach to documentation.

Users provide all components of an [FBC schema](https://olm.operatorframework.io/docs/reference/file-based-catalogs/#olm-defined-schemas), but supply only the bundle image reference for any `olm.bundle` objects. This results in a greatly simplified, smaller document.

> In a very simple example, we define an `olm.package` and an `olm.channel` composed of two `olm.bundle` objects that have an image name attribute but no other attributes/properties.

Reviewer 3: @grokspawn

> ```yaml

Reviewer 3: @grokspawn @michaelryanpeter I think we need something like here: 

```suggestion

### When is recommended for index catalog admins to adopt this veener

- If you are **not** looking to ensure standards and the content that will be distributed in your catalog
- if you are OK with Operator Authors defining channel names which make more sense for their own specific product instead of having a common nomenclature which can help cluster admins better understand the maturity of the solutions distributed in
- If you are not looking to ensure that all Operator solutions are distributed in the index following semver

This option is **not** recommended for those who are versioning the indexes and allowing Operator Authors to provide solutions to be distributed for each index version or those that are looking for common terminologies and standards on what is distributed in the catalog. If that is your case, please check the seemver veener, 

```yaml
```

Reviewer 6: I'm not sure I agree with this sentiment. If all of this was actually true, what value would there be in even creating the simple veneer in the first place? I think the purpose of these tools need to be clearly defined, but I'm skeptical that we should ever be writing documentation like this from the perspective of a superset catalog maintainer / pipeline developer that isn't directly interacting with the content of these catalogs.

Reviewer 3: Hi @kevinrizza,  

- Operators Authors produce Operators to abstract the complexities for cluster admins
- Index Catalog admins manage/built index to distribute the content on the cluster for cluster admins.
Therefore, all solutions and roles exist to address an expected result for cluster admins. 

In this way, for we choose what is the best we need to understand what problems each case scenario can solve for cluster admins and the real motivations. Otherwise, we cannot adopt A or B just because is easier to build. 

Operator Authors should NOT design solutions and make choices based on it is easy to get done, they should do something that will provide a better result for those that will consume their solutions integrated with OLM (cluster admins). 

Therefore, those that will try to validate and ensure the quality of their solution are those that (IMHO) will decide if they want to accept content made with the veener A or B. Those will also opt-in or not based on what promotes the best result for cluster admins. 

In this way, if we do not provide what are the benefits to adopt one case or another they cannot choose properly, then we will not have the best results at the end goal.

Reviewer 3: @kevinrizza I close my comment for 2 reasons:
a) It is clear to me now that the veener schema is only a helper to generate the FBC and it will not be used in another place
b) I still think that we are missing here better describes each case's motivations and what are the pros/cons of the results produced by both. However, I will do a better suggestion.

- type: olm.bundle.object
  value:

>     data: eyJhcGlWZXJzaW9uIjoiYXBpZXh0ZW5zaW9ucy5rOHMuaW8vdjEiLCJraW5kIjoiQ3VzdG9tUmVzb3VyY2VEZWZpbml0aW9uIiwibWV0YWRhdGEiOnsiYW5ub3RhdGlvbnMiOnsiY29udHJvbGxlci1nZW4ua3ViZWJ1aWxkZXIuaW8vdmVyc2lvbiI6InYwLjYuMSJ9LCJjcmVhdGlvblRpbWVzdGFtcCI6bnVsbCwibmFtZSI6ImFwcHMuZXhhbXBsZS5jb20ifSwic3BlYyI6eyJncm91cCI6ImV4YW1wbGUuY29tIiwibmFtZXMiOnsia2luZCI6IkFwcCIsImxpc3RLaW5kIjoiQXBwTGlzdCIsInBsdXJhbCI6ImFwcHMiLCJzaW5ndWxhciI6ImFwcCJ9LCJzY29wZSI6Ik5hbWVzcGFjZWQiLCJ2ZXJzaW9ucyI6W3sibmFtZSI6InYxIiwic2NoZW1hIjp7Im9wZW5BUElWM1NjaGVtYSI6eyJkZXNjcmlwdGlvbiI6IkFwcCBpcyB0aGUgU2NoZW1hIGZvciB0aGUgYXBwcyBBUEkiLCJwcm9wZXJ0aWVzIjp7ImFwaVZlcnNpb24iOnsiZGVzY3JpcHRpb24iOiJBUElWZXJzaW9uIGRlZmluZXMgdGhlIHZlcnNpb25lZCBzY2hlbWEgb2YgdGhpcyByZXByZXNlbnRhdGlvbiBvZiBhbiBvYmplY3QuIFNlcnZlcnMgc2hvdWxkIGNvbnZlcnQgcmVjb2duaXplZCBzY2hlbWFzIHRvIHRoZSBsYXRlc3QgaW50ZXJuYWwgdmFsdWUsIGFuZCBtYXkgcmVqZWN0IHVucmVjb2duaXplZCB2YWx1ZXMuIE1vcmUgaW5mbzogaHR0cHM6Ly9naXQuazhzLmlvL2NvbW11bml0eS9jb250cmlidXRvcnMvZGV2ZWwvc2lnLWFyY2hpdGVjdHVyZS9hcGktY29udmVudGlvbnMubWQjcmVzb3VyY2VzIiwidHlwZSI6InN0cmluZyJ9LCJraW5kIjp7ImRlc2NyaXB0aW9uIjoiS2luZCBpcyBhIHN0cmluZyB2YWx1ZSByZXByZXNlbnRpbmcgdGhlIFJFU1QgcmVzb3VyY2UgdGhpcyBvYmplY3QgcmVwcmVzZW50cy4gU2VydmVycyBtYXkgaW5mZXIgdGhpcyBmcm9tIHRoZSBlbmRwb2ludCB0aGUgY2xpZW50IHN1Ym1pdHMgcmVxdWVzdHMgdG8uIENhbm5vdCBiZSB1cGRhdGVkLiBJbiBDYW1lbENhc2UuIE1vcmUgaW5mbzogaHR0cHM6Ly9naXQuazhzLmlvL2NvbW11bml0eS9jb250cmlidXRvcnMvZGV2ZWwvc2lnLWFyY2hpdGVjdHVyZS9hcGktY29udmVudGlvbnMubWQjdHlwZXMta2luZHMiLCJ0eXBlIjoic3RyaW5nIn0sIm1ldGFkYXRhIjp7InR5cGUiOiJvYmplY3QifSwic3BlYyI6eyJkZXNjcmlwdGlvbiI6IkFwcFNwZWMgZGVmaW5lcyB0aGUgZGVzaXJlZCBzdGF0ZSBvZiBBcHAiLCJwcm9wZXJ0aWVzIjp7ImZvbyI6eyJkZXNjcmlwdGlvbiI6IkZvbyBpcyBhbiBleGFtcGxlIGZpZWxkIG9mIEFwcC4gRWRpdCBhcHBfdHlwZXMuZ28gdG8gcmVtb3ZlL3VwZGF0ZSIsInR5cGUiOiJzdHJpbmcifX0sInR5cGUiOiJvYmplY3QifSwic3RhdHVzIjp7ImRlc2NyaXB0aW9uIjoiQXBwU3RhdHVzIGRlZmluZXMgdGhlIG9ic2VydmVkIHN0YXRlIG9mIEFwcCIsInR5cGUiOiJvYmplY3QifX0sInR5cGUiOiJvYmplY3QifX0sInNlcnZlZCI6dHJ1ZSwic3RvcmFnZSI6dHJ1ZSwic3VicmVzb3VyY2VzIjp7InN0YXR1cyI6e319fV19LCJzdGF0dXMiOnsiYWNjZXB0ZWROYW1lcyI6eyJraW5kIjoiIiwicGx1cmFsIjoiIn0sImNvbmRpdGlvbnMiOltdLCJzdG9yZWRWZXJzaW9ucyI6W119fQ==

Author: There is a <details><summary>...</summary></details> tag group there to minimize the spill, but only in the rendered form of the doc.

relatedImages:
- image: gcr.io/kubebuilder/kube-rbac-proxy:v0.8.0
  name: ""
- image: docker.io/example/example-operator-bundle:0.2.0
  name: ""
- image: docker.io/example/example-operator:0.2.0
  name: ""
schema: olm.bundle

```
</details>

## Semver Veneer

> Since a `veneer` is identified as an input schema which is processed to generate a valid FBC, we can define a `semver veneer` as a schema which uses channel conventions to facilitate the auto-generation of channels adherent to [Semantic Versioning](https://semver.org/) (semver) guidelines. 

Reviewer 3: @grokspawn @michaelryanpeter: 

**We do not want auto-generate the channels to facilitate. We want to do that to ensure the quality of the content in the catalogue.** What motivates this feature is to provide consistent data/content across the index and not break cluster upgrades.

So, we can improve here and make more clear why the `Semver Veneer` is best for anyone that is looking for ensuring the quality of the content:

- This option has validations which avoid having upgrade graphs that do not work. That means, avoiding the scenario where misconfiguration done by Operator authors cannot submit solutions that will result in OLM users getting stuck Examples of use cases:
- invalid replaces 
- Inconsistent channels names which have no meaning for cluster admins 
- Ensure that channels and Operators Bundle versions follow up the semver convention which allows cluster admins know exactly what are the versions that then can safely upgrade because are only patch releases and those ones which are versions that they MUST plan beforehand to upgrade because are breaking changes
- Another advantage here is to ensure that the whole catalogue provides the same terminology for the channels which makes it easier for cluster admins to know what is the maturity of the solutions published under the channels instead of each Operator Author choosing a term that might only make sense for them. 
- Encourage and promote standards across the catalog 

Therefore, we should recommend its adoption for those Index Admins which are versioning indexes to allow Operator Authors such as `myregistry/myindex:1.0` and `myregistry/myindex:20`

/hold

I am holding this one because this part IMO is very important. We need really be able to clarify what are the problems that these options address properly here.

Reviewer 3: ```suggestion
Since a `veneer` is identified as an input schema which is processed to generate a valid FBC, we can define a `semver veneer` as a schema which uses channel conventions to ensure standards and the quality of the content distribute in the index catalog. Therefore, this veener encourages the adoption of seemver and provides channel auto-generation adherent to [Semantic Versioning](https://semver.org/) (semver) guidelines to facilitate the consumption of what is distributed within and its consumption for cluster admins.  

### When is recommended for index catalog admins to adopt this veener

- If they are looking to ensure that all Operator solutions distributed on the index will follow up semver
- If they are looking to ensure consistent terminology across the index catalogue intend to help cluster admins better and more intuitively understand the maturity level of what is distributed in each channel. 
- If you are versioning the indexes and allowing Operator authors to provide different packages schemas per index version this option can avoid production disruptions such as when cluster admins must manually change the subscription because after a cluster or index upgrades the channel where they are subscribed no longer exist
- If you are looking to facilitate for cluster admins are able to properly know what are the versions that they can safely upgrade because for example just a patch release or when they must better plan these upgrades in production because the next version is a MAJOR bump which contains braking changes and can affect their workloads. 

```

Author: I don't agree with you here.  The operator author benefits from the reduced labor.  Our catalog benefits from the consistency.  All the author cares about w.r.t. the catalog is that it's easy to get their operator there, and for users to pull it from there.

Reviewer 3: HI @grokspawn, 

> I don't agree with you here. The operator author benefits from the reduced labor.

You can add it as an advantage but that is not what motivates the veener.  
The motivation for looking for generating the channels is not to make it easier for the Operator authors at all. It is about ensuring consistency for those who will consume the solutions.

We have a huge problem with OCP cluster upgrades because when cluster admins upgrade their cluster they get stuck when the new index version does not exist in the channel where they are subscribed. The same bad user experience can happen if we create an index per k8s versions releases. (It is a very common case scenario that the best practices naming convention try to sort out. we shared it in the f2f as well. )

Who chooses the veener? Are the Operator Authors or the Index Admin? 

Therefore, the index admin will make the choice according to what is best to address their needs in order to make things better for those who consume the catalog itself which is the cluster admin.

Author: Only the operator author will interact with veneers.  The rest of the pipeline deals only with FBC.  Therefore, the operator author selects the veneer that best suits them (or forks, implements their own, etc.).
  
I can totally foresee that -- if we do it right -- index owners will come to prefer veneer-generated FBC instead of hand-generated FBC.  I guess that's where I see the benefits of an incentivized opt-in approach (based on award/achievement?), since then index owners can incentivize  a "this tall to ride" approach.  

For example, if the lead time for index inclusion is typically 1 week, the index owner can offer to shorten it to a couple of days if a veneer was used to generate an operator's FBC (due to a higher confidence in the consistency and durability of the resulting FBC). 

I believe that a prescriptive approach will be detrimental to our contributing authors.  I have had a look at the marketplace catalogs, and they are considerably disparate from what we propose,.  Therefore, any required compliance would introduce friction as it imposes costs.  

Since we are in early alpha, we are working to interact with our contributing authors to define the veneers which do the best job of capturing their use-cases.

> [**DISCLAIMER:** since version build metadata [MUST be ignored when determining version precedence](https://semver.org/#spec-item-10) when using semver, it cannot be used in any bundle included in the `semver veneer` and will result in a fatal error.]

Author: This is just documenting the current implementation, see for e.g. https://github.com/operator-framework/operator-registry/pull/948

> ### Schema Goals

Reviewer 3: What about we add all that is from each veener as a subsection inside instead of creating this other topic?
Example

## Veener A

### How does it work?
### When should I use it
### Schemas 

## Veener B

### How does it work?
### When should I use it
### Schemas

Reviewer 2: I think there is sense in tweaking the information architecture. Personally, I find headers that are questions distracting in docs that aren't FAQs.

WDYT of something like:

- Basic veneer use-cases
- Basic veneer workflow
- Building a file-based catalog with the basic veneer (for the procedure)
- Example basic veneer
- Basic veneer reference (or schema)

Reviewer 2: This approach to the headings would would also help with SEO

The `semver veneer` must have:
- terse grammar to minimize creation/maintenance effort
- deterministic output
- simple channel promotion for maturing bundles
- demonstration of a common type of channel maturity model
- minor-version (Y-stream), major-version (X-stream) versioning optionality

The resulting FBC must clearly indicate how generated channels relate to veneer entities.

> ### Schema Anatomy

Reviewer 3: WDYT about the titles be more friendly for the users like: When should I use it? How does it work? Examples and etc ..

  - Image: quay.io/foo/olm:testoperator.v0.2.2
  - Image: quay.io/foo/olm:testoperator.v0.3.0
  - Image: quay.io/foo/olm:testoperator.v1.0.0
  - Image: quay.io/foo/olm:testoperator.v1.0.1
  - Image: quay.io/foo/olm:testoperator.v1.1.0
Fast:
  Bundles:
  - Image: quay.io/foo/olm:testoperator.v0.2.1
  - Image: quay.io/foo/olm:testoperator.v0.2.2
  - Image: quay.io/foo/olm:testoperator.v0.3.0
  - Image: quay.io/foo/olm:testoperator.v1.0.1
  - Image: quay.io/foo/olm:testoperator.v1.1.0
Stable:
  Bundles:
  - Image: quay.io/foo/olm:testoperator.v1.0.1
```
In this example, `Candidate` has the entire version range of bundles,  `Fast` has a mix of older and more-recent versions, and `Stable` channel only has a single published entry. 

### CLI Tool Usage
```
% ./bin/opm alpha render-veneer semver -h
Generate a file-based catalog from a single 'semver veneer' file

Usage:
  opm alpha render-veneer semver <filename> [flags]

Flags:
  -h, --help            help for semver
  -o, --output string   Output format (json|yaml|mermaid) (default "json")

Global Flags:
      --skip-tls-verify   skip TLS certificate verification for container image registries while pulling bundles

>       --use-http          use plain HTTP for container image registries while pulling bundles

Reviewer 2: Downstream, we would break each command and the output up into separate codeblocks. It helps with copying commands and separating the example output. 

Also, WDYT of putting the flags in a table instead of printing the CLI output? I think it would help with scanability in the doc and make for a reference that is easier to read.

---
title: "Veneers"
linkTitle: "Veneers (alpha)"
weight: 5
date: 2022-06-30
---

>Note: `veneers` are **ALPHA** functionality and may adopt breaking changes

## Concept and Design

File-Based Catalogs (FBC) are a major improvement to the imperative update graph approaches of previous versions. FBCs give operator authors a [declarative and deterministic approach to defining their update graph](https://olm.operatorframework.io/docs/concepts/olm-architecture/operator-catalog/creating-an-update-graph/). However, FBCs can get complex, especially as the number of releases and dependencies scale. We introduce the concept of a `veneer` as an approach to simplifying interacting with FBCs.

> 

Reviewer 3: ```suggestion

By using the `veneer` and the [opm](https://github.com/operator-framework/operator-registry) tool you can abstract the complexities to generate the content for your Operator package in the file-based-format.
```

Author: It's true that we currently support two veneer types, but a veneer is a simplification, not just the opm-supported one.  

That's why I broke out the veneer concept from the 'supported by opm' concept.  

If an author wants to use another approach which results in valid FBC, it would still be a valid veneer.

Reviewer 3: @grokspawn can we try  something like 

 `veneer` can be supported and provided by other tools. In this document you will check how the `veneers` which are supported by [opm][opm-tool] tool can abstract the complexities to generate the content for your Operator package in the file-based-format.

Reviewer 3: Maybe before `The veneers supported by `opm` are:`. would fit better

Reviewer 3: Sure, I think we could just clarifies that a little bit more : https://github.com/operator-framework/olm-docs/pull/245#discussion_r927078092

In this context, there are two components to every `veneer`:
1. An arbitrary API
2. An executable which processes #1 and produces a valid FBC.

> The veneers supported by `opm` are:

Reviewer 3: ```suggestion
The veneers supported by [opm](https://github.com/operator-framework/operator-registry)  are:
```

Reviewer 3: @grokspawn @michaelryanpeter can we link here ?

- type: olm.bundle.object
  value:
    data: eyJhcGlWZXJzaW9uIjoicmJhYy5hdXRob3JpemF0aW9uLms4cy5pby92MSIsImtpbmQiOiJDbHVzdGVyUm9sZSIsIm1ldGFkYXRhIjp7ImNyZWF0aW9uVGltZXN0YW1wIjpudWxsLCJuYW1lIjoiZXhhbXBsZS1vcGVyYXRvci1tZXRyaWNzLXJlYWRlciJ9LCJydWxlcyI6W3sibm9uUmVzb3VyY2VVUkxzIjpbIi9tZXRyaWNzIl0sInZlcmJzIjpbImdldCJdfV19
- type: olm.bundle.object
  value:
    data: eyJhcGlWZXJzaW ... [snip] ... jEuMCJ9fQ==
- type: olm.bundle.object
  value:
    data: eyJhcGlWZXJzaW ... [snip] ... jEuMCJ9fQ==
relatedImages:
- image: gcr.io/kubebuilder/kube-rbac-proxy:v0.8.0
  name: ""
- image: docker.io/example/example-operator-bundle:0.2.0
  name: ""
- image: docker.io/example/example-operator:0.2.0
  name: ""
schema: olm.bundle

```
</details>

## Semver Veneer

> Since a `veneer` is identified as an input schema which is processed to generate a valid FBC, we can define a `semver veneer` as a schema which uses channel conventions to facilitate the auto-generation of channels adherent to [Semantic Versioning](https://semver.org/) (semver) guidelines. This approach may be attractive to operator authors who are defining a new upgrade graph, or are already close enough to the veneer conventions to be able to adopt it (i.e. already use semver, only promote a single channel, etc.).

Reviewer 3: ```suggestion
Since a `veneer` is identified as an input schema which is processed to generate a valid FBC, we can define a `semver veneer` as a schema which uses [channel naming conventions](https://olm.operatorframework.io/docs/best-practices/channel-naming/) to facilitate the auto-generation of channels adherent to [Semantic Versioning](https://semver.org/) (semver) guidelines. This approach may be attractive to operator authors who are defining a new upgrade graph, or are already close enough to the veneer conventions to be able to adopt it (i.e. already use semver, only promote a single channel, etc.).
```

Reviewer 3: what about we link here?

---
title: "Veneers"
linkTitle: "Veneers (alpha)"
weight: 5
date: 2022-06-30
---

>Note: `veneers` are **ALPHA** functionality and may adopt breaking changes

## Concept and Design

> File-Based Catalogs (FBC) are a major improvement to the imperative update graph approaches of previous versions. FBCs give operator authors a [declarative and deterministic approach to defining their update graph](https://olm.operatorframework.io/docs/concepts/olm-architecture/operator-catalog/creating-an-update-graph/). However, FBCs can get complex, especially as the number of releases and dependencies scale. We introduce the concept of a `veneer` as an approach to simplifying interacting with FBCs.

Reviewer 7: ```suggestion
File-Based Catalogs (FBC) are a major improvement to the imperative update graph approaches of previous versions. FBCs give operator authors a [declarative and deterministic approach to defining their update graph](https://olm.operatorframework.io/docs/concepts/olm-architecture/operator-catalog/creating-an-update-graph/). However, FBCs can get complex, especially as the number of releases and dependencies scale. We introduce the concept of a `veneer` as an approach to simplifying interacting with, ang generating valid, FBCs.
```

In this context, there are two components to every `veneer`:

> 1. An arbitrary API

Reviewer 7: It feels a bit unclear what is an arbitrary API in this context. Should we provide an example?

Reviewer 7: Maybe, for instance an input yaml file that respects an arbitrary schema

Author: I was deliberately trying to avoid being opinionated here.  The critical aspect is that whatever it is, in #2 it must generate a valid FBC. 
For e.g. all these are valid `veneers` when they output valid FBC:
- XSLT which processes internal cataloging system XML
- build process which knows how to insert a new build
- script to take a text list of operators in explicit order and generate the channel edges for a template

"A veneer is anything that can be converted through an arbitrary process into a valid FBC" didn't have the same gravitas. 

I'm open to anything that doesn't narrow the definition too much, since the critical aspect is the end-state, not the start-state.

> 2. An executable which processes #1 and produces a valid FBC.

Reviewer 3: I would prefer something like: 

“veneer is API, you can extend and create your own veeners. Here we will show the veeners supported by OPM and how they can help you out abstract the complex to generate your Operator package metadata in the FBC format. Note: the veener syntax can only help you generate FBC manifests”  

wdyt?
